hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/LocalProcessLockProvider.java

```diff
@@ -0,0 +1,117 @@
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hudi.common.config.LockConfiguration;
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.lock.LockProvider;
+import org.apache.hudi.common.lock.LockState;
+import org.apache.hudi.common.util.StringUtils;
+import org.apache.hudi.exception.HoodieLockException;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Local process level lock. This {@link LockProvider} implementation is to
+ * guard table from concurrent operations happening in the local JVM process.
+ * <p>
+ * Note: This Lock provider implementation doesn't allow lock reentrancy.
+ * Attempting to reacquire the lock from the same thread will throw
+ * HoodieLockException. Threads other than the current lock owner, will
+ * block on lock() and return false on tryLock().
+ */
+public class LocalProcessLockProvider implements LockProvider<ReentrantReadWriteLock> {
+
+  private static final Logger LOG = LogManager.getLogger(ZookeeperBasedLockProvider.class);
+  private static final ReentrantReadWriteLock LOCK = new ReentrantReadWriteLock();
+  private final long maxWaitTimeMillis;
+
+  public LocalProcessLockProvider(final LockConfiguration lockConfiguration, final Configuration conf) {
+    TypedProperties typedProperties = lockConfiguration.getConfig();
+    maxWaitTimeMillis = (typedProperties.containsKey(LockConfiguration.LOCK_ACQUIRE_WAIT_TIMEOUT_MS_PROP_KEY)
+        ? lockConfiguration.getConfig().getLong(LockConfiguration.LOCK_ACQUIRE_WAIT_TIMEOUT_MS_PROP_KEY) :
+        LockConfiguration.DEFAULT_ACQUIRE_LOCK_WAIT_TIMEOUT_MS);
+  }
+
+  @Override
+  public void lock() {
+    LOG.info(getLogMessage(LockState.ACQUIRING));
+    if (LOCK.isWriteLockedByCurrentThread()) {
+      throw new HoodieLockException(getLogMessage(LockState.ALREADY_ACQUIRED));
+    }
+    LOCK.writeLock().lock();
+    LOG.info(getLogMessage(LockState.ACQUIRED));
+  }
+
+  @Override
+  public boolean tryLock() {
+    return tryLock(maxWaitTimeMillis, TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public boolean tryLock(long time, @NotNull TimeUnit unit) {
+    LOG.info(getLogMessage(LockState.ACQUIRING));
+    if (LOCK.isWriteLockedByCurrentThread()) {
+      throw new HoodieLockException(getLogMessage(LockState.ALREADY_ACQUIRED));
+    }
+
+    boolean isLockAcquired;
+    try {
+      isLockAcquired = LOCK.writeLock().tryLock(time, unit);
+    } catch (InterruptedException e) {
+      throw new HoodieLockException(getLogMessage(LockState.FAILED_TO_ACQUIRE));
+    }
+
+    LOG.info(getLogMessage(isLockAcquired ? LockState.ACQUIRED : LockState.FAILED_TO_ACQUIRE));
+    return isLockAcquired;
+  }
+
+  @Override
+  public void unlock() {
+    LOG.info(getLogMessage(LockState.RELEASING));
+    if (!LOCK.isWriteLockedByCurrentThread()) {
```

Comment:
`unlock` will throw if we're trying to release from the thread that is not holding the lock -- if we want to throw our own exception we should instead wrap unlock in try-catch and convert (since this conditional isn't guaranteeing us that it will not throw anyway)

Reply:
right, letting the unlock take care of the validation is the right idea. fixed.